### PR TITLE
Allow `<` and `>` in strings when using `--no-stdlib`

### PR DESCRIPTION
**Pull request by Morriar**
### Motivation

Running the following command gives no error which is as expected:

~~~bash
$ sorbet -e "puts 'Hello > World'"
~~~

While running the same command with `--no-stdlib` gives:

~~~bash
$ sorbet --no-stdlib -e "puts 'Hello > World'"
Exception::raise(): core/GlobalState.cc:723 enforced condition nm.find(">") == string::npos has failed: (no message provided)

Backtrace:
sorbet::Exception::enforce_handler(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, int) (in librealmain.so) + 106
sorbet::core::GlobalState::enterString(std::__1::basic_string_view<char, std::__1::char_traits<char> >) (in libcore.so) (GlobalState.cc:720)
sorbet::core::GlobalState::enterNameUTF8(std::__1::basic_string_view<char, std::__1::char_traits<char> >) (in libcore.so) (GlobalState.cc:794)
sorbet::parser::Builder::Impl::string(ruby_parser::token const*) (in libparser.so) + 107
(anonymous namespace)::string_(void const*, ruby_parser::token const*) (in libparser.so) + 60
ruby_parser::bison::typedruby25::parser::parse() (in libparser.so) + 82416
ruby_parser::typedruby25::parse(void const*) (in libparser.so) + 50
sorbet::parser::Builder::build(ruby_parser::base_driver*) (in libparser.so) + 113
sorbet::parser::Parser::run(sorbet::core::GlobalState&, sorbet::core::FileRef) (in libparser.so) + 290
sorbet::realmain::pipeline::runParser(sorbet::core::GlobalState&, sorbet::core::FileRef, sorbet::realmain::options::Printers const&) (in libpipeline.so) + 507
sorbet::realmain::pipeline::indexOneWithPlugins(sorbet::realmain::options::Options const&, sorbet::core::GlobalState&, sorbet::core::FileRef, std::__1::unique_ptr<sorbet::KeyValueStore, std::__1::default_delete<sorbet::KeyValueStore> >&) (in libpipeline.so) + 1005
sorbet::realmain::pipeline::index(std::__1::unique_ptr<sorbet::core::GlobalState, std::__1::default_delete<sorbet::core::GlobalState> >&, std::__1::vector<sorbet::core::FileRef, std::__1::allocator<sorbet::core::FileRef> >, sorbet::realmain::options::Options const&, sorbet::WorkerPool&, std::__1::unique_ptr<sorbet::KeyValueStore, std::__1::default_delete<sorbet::KeyValueStore> >&) (in libpipeline.so) + 476
sorbet::realmain::realmain(int, char**) (in librealmain.so) + 6186
main (in sorbet) (main.cc:7)
start (in libdyld.dylib) + 1

-e: Exception parsing file: -e (backtrace is above) https://srb.help/1001
Errors: 1
~~~

Instead of:

~~~
-e:1: Method puts does not exist on T.class_of(<root>) https://srb.help/7003
     1 |puts "Hello > World"
        ^^^^^^^^^^^^^^^^^^^^
Errors: 1
~~~

See added tests in `test/cli/no-stdlib`.

### Approach

I tracked down the problem to `GlobalState::enterString` which is called by `enterNameUTF8` to build strings (`Builder.cc`): 

https://github.com/sorbet/sorbet/blob/ea80b0490f18f61f91170adefc682ba510556b0e/core/GlobalState.cc#L720-L725

This was introduced by commit fb09ecd887bd7703550e0648298c516723bc4b76 (cc. @ptarjan) and the commit message says it's to prevent `>` and `<` in *names* but apparently also applies to *strings*.

Reverting the commit makes Sorbet behave correctly and doesn't break any test so my question is: "Why do we need this check?". It seems to be related to some internal things Sorbet might be doing since the parser will already raise an error if a name contains `>` or `<`.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
@Morriar would the solution here not be to mark global state as initialized if the `--no-stdlib` flag is passed?

File: core/GlobalState.cc [hunk 1]
```diff
@@ -717,12 +715,6 @@ ArgInfo &GlobalState::enterMethodArgumentSymbol(Loc loc, SymbolRef owner, NameRe
 }
 
 string_view GlobalState::enterString(string_view nm) {
-    DEBUG_ONLY(if (isInitialized) {
-        if (nm != "<" && nm != "<<" && nm != "<=" && nm != "<=>" && nm != ">" && nm != ">>" && nm != ">=") {
```
Comment (DarkDimius) [inline core/GlobalState.cc]: this test has found a lot of subtle bugs and I'd like to retain it if possible.

Comment (DarkDimius) [reply core/GlobalState.cc]: Shouldn't this `if` have filtered it out and allowed it?

Comment (Morriar) [reply core/GlobalState.cc]: @DarkDimius can you tell us more about what it prevents and why it's done while parsing strings?

I'm happy to provide a better/cleaner fix but I'm not sure to understand what is done here and why?

Comment (DarkDimius) [reply core/GlobalState.cc]: We have a lot of internal names of form `<something>` that's chosen with `<` and `>` as you can't make this into a valid ruby identifier without suffering.
We want to make sure we don't round-trip through `string` for those names.

**Comment by Morriar** [comment]
@ptarjan I pushed another version for this:

* Rename and document the `isInitialized` -> `ensureCleanStrings` flag so it explains a bit more what it does
* Enable the flag before loading the payload from source then disable it after the payload is loaded

If `--no-stdlib` is set there is no payload loading required so no checks to do.
If `nameTablePayload != null`, `GlobalState::initEmpty` was not called so I guess no check to perform here either.

File: payload/payload.cc [hunk 2]
```diff
@@ -52,6 +53,7 @@ void createInitialGlobalState(unique_ptr<core::GlobalState> &gs, const realmain:
         auto workers = WorkerPool::create(emptyOpts.threads, gs->tracer());
         auto indexed = realmain::pipeline::index(gs, payloadFiles, emptyOpts, *workers, kvstore);
         realmain::pipeline::resolve(gs, move(indexed), emptyOpts, *workers); // result is thrown away
+        gs->ensureCleanStrings = false;
```
Comment (DarkDimius) [inline payload/payload.cc]: this only checks it during initial global state construction, right?
I don't think this covers usecases that we want checked. We want to run Sorbet test suite on inputs from test suite with it enabled.

Comment (Morriar) [reply payload/payload.cc]: @DarkDimius: Yes it only checks payload loading.

> We want to run Sorbet test suite on inputs from test suite with it enabled.

That's the part I didn't understood, now I see the problem, sorry!

**Comment by Morriar** [comment]
@DarkDimius and @ptarjan: I tried to understand when this check was actually enabled.

For the check to happen, the `GlobalState::isInitialized` attribute must be set to `true` (https://github.com/sorbet/sorbet/blob/master/core/GlobalState.cc#L720-L721). For this to happen the `GlobalState::initEmpty()` method must be called which means there are only two code paths where this check is activated:

* When loading no payload (with the `--no-stdlib` option)
* When loading the payload from source (`read_global_state.source`)

When loading the payload from a KV store (`read_global_state.kvstore`) or directly from it’s binary representation (`read_global_state.binary`), no check is enabled. So, **over 4 ways to initialize Sorbet, only two have the check enabled**. 

Looking at how the tests are ran, **I don’t think this check was actually enabled on the test inputs**. The only concrete way to enable it, was to run it with `--stdlib` hence the existence of this PR.

Let’s try to clarify this a bit more...

Sorbet is loading code in two situations:

1. Loading the stdlib payload (source, kvstore, binary or none with `--nostdlib`)
2. Loading the actual Ruby code to typecheck

**Do we want the checks to be run in both these situations?**

For 1, enabling the checks while loading payload code is already proposed in this PR.

If we also want 2, here two more questions:

What about these two tests which contain the forbidden tokens while being valid Ruby?

* https://github.com/sorbet/sorbet/blob/master/test/testdata/parser/whitequark/parse_lvar_injecting_match.rb
* https://github.com/sorbet/sorbet/blob/master/test/testdata/parser/whitequark/parse_non_lvar_injecting_match.rb

What about the DSL for Minitest which generate names with `<` and `>`?

* https://github.com/sorbet/sorbet/blob/master/dsl/Minitest.cc#L73
* https://github.com/sorbet/sorbet/blob/master/dsl/Minitest.cc#L77

**Comment by DarkDimius** [comment]
Agreed, I think the property that we wanted to enforce indeed isn't checked now.
Sorry, Alexandre for confusion and thank you for pushing through.

**Pull request merged**